src/data/W/basic.lean

```diff
@@ -37,7 +37,73 @@ instance : inhabited (W_type (λ (_ : unit), empty)) :=
 
 namespace W_type
 
-variables {α : Type*} {β : α → Type*} [Π a : α, fintype (β a)]
+variables {α : Type*} {β : α → Type*}
+
+/-- The canonical map to the corresponding sigma type, returning the label of a node as an
+  element `a` of `α`, and the children of the node as a function `β a → W_type β`. -/
+def to_sigma : W_type β → Σ a : α, β a → W_type β
+| ⟨a, f⟩ := ⟨a, f⟩
+
+/-- The canonical map from the sigma type into a `W_type`. Given a node `a : α`, and
+  its children as a function `β a → W_type β`, return the corresponding tree. -/
+def of_sigma : (Σ a : α, β a → W_type β) → W_type β
+| ⟨a, f⟩ := W_type.mk a f
+
+@[simp] lemma of_sigma_to_sigma : Π (w : W_type β),
+  of_sigma (to_sigma w) = w
+| ⟨a, f⟩ := rfl
+
+@[simp] lemma to_sigma_of_sigma : Π (s : Σ a : α, β a → W_type β),
+  to_sigma (of_sigma s) = s
+| ⟨a, f⟩ := rfl
+
+variable (β)
+
+/-- The canonical bijection with the sigma type, showing that `W_type` is a fixed point of
+  the polynomial `Σ a : α, β a → W_type β`.  -/
+def equiv_sigma : W_type β ≃ Σ a : α, β a → W_type β :=
+{ to_fun := to_sigma,
+  inv_fun := of_sigma,
+  left_inv := of_sigma_to_sigma,
+  right_inv := to_sigma_of_sigma }
+
+variable {β}
+
+/-- The canonical map from `W_type β` into any type `γ` given a map `(Σ a : α, β a → γ) → γ`. -/
+def to_type (γ : Type*) (fγ : (Σ a : α, β a → γ) → γ) : W_type β → γ
```

Comment:
Am I right that this is an alternative induction principle?

Reply:
Yes, this could be called `W.elim` perhaps?